File: kindle-clippings.go
package main

import (
	"bytes"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"path"
	"regexp"
	"sort"
	"strings"
)

var metadataReg = regexp.MustCompile(`\-\s+Your\s(.*?)\s+on\s+(.*?)\s*\|.*`)

func panicIfErr(err error) {
	if err != nil {
		panic(err)
	}
}

func main() {
	var toFile string

	flag.StringVar(&toFile, "to", "", "Append to file")
	flag.Parse()

	for _, filename := range flag.Args() {
		if !strings.HasSuffix(filename, ".txt") {
			filename = path.Join(filename, "documents/My Clippings.txt")
		}
		bytes, err := ioutil.ReadFile(filename)
		panicIfErr(err)

		res := parseKindleNotes(string(bytes))
		fmt.Println(res)

		if len(toFile) > 0 {
			f, err := os.OpenFile(toFile, os.O_APPEND|os.O_WRONLY, 0600)
			panicIfErr(err)
			_, err = f.WriteString(res)
			panicIfErr(err)
			_ = f.Close()

			fmt.Printf("Appended to %s\n", toFile)
			fmt.Printf("Empty %s [y/N]?", filename)
			var answer string
			fmt.Scanf("%s", &answer)
			if answer == "y" {
				fmt.Println("Emptying " + filename)
				err = ioutil.WriteFile(filename, []byte{}, 777)
				panicIfErr(err)
			}
		}
	}
}

func parseKindleNotes(str string) string {
	var keys []string
	var texts = map[string][]string{}

	var res bytes.Buffer
	var unknown bytes.Buffer
	parts := strings.Split(str, "==========")
	for _, part := range parts {
		part = strings.TrimLeft(part, " \n\r\t")
		lines := strings.Split(part, "\n")
		if len(lines) < 3 {
			unknown.WriteString(part)
		} else {
			title := lines[0]
			metadata := lines[1]
			text := strings.TrimSpace(strings.Join(lines[2:], "\n"))
			//fmt.Println("title:", title)
			//fmt.Println("text:", text)

			if metadataReg.MatchString(metadata) {
				all := metadataReg.FindAllStringSubmatch(metadata, -1)
				typ := all[0][1]
				loc := all[0][2]

				if typ == "Note" {
					text = "Note: " + text
				}
				if typ == "Highlight" || typ == "Note" {
					typ = "Txt"
				}

				//fmt.Printf("type=%s location=%s\n", typ, loc)

				key := fmt.Sprintf("%s :: %s", strings.TrimSpace(title), strings.TrimSpace(typ))
				if _, found := texts[key]; !found {
					keys = append(keys, key)
					texts[key] = []string{}
				}
				texts[key] = append(texts[key], text)
			} else {
				unknown.WriteString(part)
			}
		}
	}

	sort.Strings(keys)

	for _, key := range keys {
		txt := strings.TrimSpace(strings.Join(texts[key], "\n"))
		if len(txt) == 0 {
			continue
		}
		res.WriteString("# " + key + "\n\n")
		res.WriteString(txt)
		res.WriteString("\n\n")
	}

	if unknown.Len() > 0 {
		res.WriteString("# Unknown\n\n")
		res.WriteString(unknown.String())
	}

	return res.String()
}

File: kindle-clippings.py
#!/usr/bin/env python3

import subprocess
import datetime
import os
import os.path as path
import re
import shutil

from typing import *


mount_points = "/Volumes"
now_timestamp_str = datetime.datetime.now().strftime('%Y-%m-%dT%H:%M')

def find_clipping_files() -> List[str]:
	files: List[str] = []
	file_name = "documents/My Clippings.txt"
	for dir_name in os.listdir(mount_points):
		fn = path.join(mount_points, dir_name, file_name)
		if path.exists(fn):
			files.append(fn)
	return files

def parse_kindle_notes(file_name: str) -> str:
	keys: List[str] = []
	texts: Dict[str, List[str]] = {}
	unknown = ""

	with open(file_name) as f:
		contents = f.read()
	parts = contents.split("==========")
	for part in parts:
		part = part.strip()
		lines = part.split("\n")
		if len(lines) < 3:
			# unknown += part + "\n"
			pass
		else:
			title = lines[0]
			metadata = lines[1]
			text = "\n".join(lines[2:])
			m = re.match('\-\s+Your\s(.*?)\s+on\s+(.*?)\s*\|.*', metadata)
			if m:
				typ = m.group(1)
				if typ == "Note":
					text = "Note: " + text
				if typ == "Highlight" or typ == "Note":
					typ = "Txt"
				key = f"{title} :: {typ}"
				if not key in texts:
					keys.append(key)
					texts[key] = []
				texts[key].append(text)
			else:
				unknown += part + "\n"

	res = f"\n# Imported {now_timestamp_str}\n"
	keys.sort()
	for key in keys:
		txt = "\n".join(texts[key])
		res += f"\n## {key}\n"
		res += txt + "\n"

	if unknown:
		res += "\n## Unknown\n"
		res += unknown + "\n"

	return res

def main() -> None:
	for file in find_clipping_files():
		contents = parse_kindle_notes(file)
		target_file = "kindle_clippings.md"
		print(contents)
		if input(f"Append to {target_file}? [yn]") == "y":
			with open(target_file, "a") as f:
				f.write(contents)
			if input(f"Clear {file}? [yn]") == "y":
				with open(file, "w") as f:
					f.write("")
				backup_file = f"kindle_clippings_backup_{now_timestamp_str}.txt"
				shutil.copy2(file, backup_file)
				print(f"Backup {file} -> {backup_file}")

main()